### PR TITLE
Meeting is over sooner

### PR DESCRIPTION
**Pull request by jtkpiotr**
@michal-konarski @lkudyba @quirk0o @mhib  @nowNick @adamzima @s1mplex

### CODE REVIEW

File: app/models/google_calendar/event_editor.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    def event_finished?
+      event.end.date_time < current_time
+    end
+
+    def current_time
+      DateTime.now
+    end
+
+    attr_accessor :event_id, :calendar_service, :conference_room_email, :event
```
Comment (mjk-old) [inline app/models/google_calendar/event_editor.rb]: Is there a convention to put this at the very end? I'd put it right below `private`.

Additionally, do you ever use `event_id` as an instance variable?

File: app/models/google_calendar/event_editor.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
+module GoogleCalendar
+  class EventEditor
+    def initialize(credentials)
+      @credentials = credentials
+      @calendar_service = GoogleCalendar::Client.new(credentials).calendar_service
+    end
+
+    def finish(conference_room_id, event_id)
+      @conference_room_email = ConferenceRoom.find(conference_room_id).email
```
Comment (mjk-old) [inline app/models/google_calendar/event_editor.rb]: I don't think `@conference_room_email` and `@event` should be instance variables. They are not properties of the class, cause `EventEditior` allows to edit any events, not only these particular. After invoking `finish` they no longer make sense as a part of this class. You should pass them around as method arguments.

File: app/models/google_calendar/event_editor.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      calendar_service.update_event(conference_room_email, event.id, event)
+    end
+
+    def finish_event
+      return if event_not_started? || event_finished?
+      event.end.date_time = current_time
+      update_event
+    end
+
+    def event_not_started?
```
Comment (mjk-old) [inline app/models/google_calendar/event_editor.rb]: Move this logic to `event` itself. Additionally, change these two methods to `in_progress?` and invert the condition. 

Please notice that now you check these two conditions against two different DataTime objects. It's not a problem here, but in general you shouldn't do this. It more robust to have these two conditions merged into one.

File: app/models/google_calendar/event_editor.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    def get_event(event_id)
+      calendar_service.get_event(conference_room_email, event_id)
+    end
+
+    def update_event
+      calendar_service.update_event(conference_room_email, event.id, event)
+    end
+
+    def finish_event
+      return if event_not_started? || event_finished?
```
Comment (mjk-old) [inline app/models/google_calendar/event_editor.rb]: Have you considered how would it behave when two users hit this method for the same event at the same time simultaneously?

File: app/controllers/events_controller.rb [hunk 5]
```diff
@@ -60,22 +60,28 @@ def destroy
...
-    event_id = confirmation_params[:event_id]
-    Event.confirm_or_create(conference_room_id, event_id)
+    Event.confirm_or_create(edit_event_params[:conference_room_id], edit_event_params[:event_id])
+    head :ok
+  end
+
+  def finish
+    google_event_client.finish(edit_event_params[:conference_room_id], edit_event_params[:event_id])
+    head :ok
+  rescue ActiveRecord::RecordNotFound
```
Comment (adamzima) [inline app/controllers/events_controller.rb]: You already had implemented rescue from RecordNotFound globally

File: app/models/google_calendar/event_wrapper/event.rb [hunk 6]
```diff
@@ -25,6 +25,20 @@ def valid?
...
 
+      def all_day?
+        start.date.present?
+      end
+
+      def in_progress?
+        current_time >= start_time.date_time && current_time <= end_time.date_time
+      end
+
+      def finish
```
Comment (adamzima) [inline app/models/google_calendar/event_wrapper/event.rb:36]: Since you modify an object here, I would name it `finish!`

File: config/routes.rb [hunk 7]
```diff
@@ -8,6 +8,7 @@
   post 'notify/:conference_room_id' => 'notification#receive', as: :notifications
   get 'conference_rooms/:conference_room_id/events' => 'events#room_index'
   post 'conference_rooms/:conference_room_id/events/:event_id/confirm' => 'events#confirm', as: :confirmation
+  post 'conference_rooms/:conference_room_id/events/:event_id/finish' => 'events#finish', as: :finish
```
Comment (adamzima) [inline config/routes.rb]: Why aren't you using resources?

File: app/models/google_calendar/event_wrapper/event.rb (hunk 6 above)
Comment (mjk-old) [reply app/models/google_calendar/event_wrapper/event.rb:36]: But then you'll make Prima Donna code smell ;) https://github.com/troessner/reek/blob/master/docs/Prima-Donna-Method.md

**Comment by mjk-old** [comment]
👍 ok

**Pull request merged**